# Lasso Regression Without Sklearn Packages

## Importing required Libraries

In [51]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

## Reading the dataset

In [52]:
# Importing DataSet and take a look at Data
data = pd.read_csv("boston_train.csv")

In [53]:
# checking first five rows
data.head()

ID     crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0   1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1   2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2   4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
3   5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
4   7  0.08829  12.5   7.87     0  0.524  6.012  66.6  5.5605    5  311   

   ptratio   black  lstat  medv  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     18.7  394.63   2.94  33.4  
3     18.7  396.90   5.33  36.2  
4     15.2  395.60  12.43  22.9

In [54]:
#checking both no null values and columns dtype
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 333 entries, 0 to 332
Data columns (total 15 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   ID       333 non-null    int64  
 1   crim     333 non-null    float64
 2   zn       333 non-null    float64
 3   indus    333 non-null    float64
 4   chas     333 non-null    int64  
 5   nox      333 non-null    float64
 6   rm       333 non-null    float64
 7   age      333 non-null    float64
 8   dis      333 non-null    float64
 9   rad      333 non-null    int64  
 10  tax      333 non-null    int64  
 11  ptratio  333 non-null    float64
 12  black    333 non-null    float64
 13  lstat    333 non-null    float64
 14  medv     333 non-null    float64
dtypes: float64(11), int64(4)
memory usage: 39.1 KB


In [55]:
#checking for the outliers
data.describe()

ID        crim          zn       indus        chas         nox  \
count  333.000000  333.000000  333.000000  333.000000  333.000000  333.000000   
mean   250.951952    3.360341   10.689189   11.293483    0.060060    0.557144   
std    147.859438    7.352272   22.674762    6.998123    0.237956    0.114955   
min      1.000000    0.006320    0.000000    0.740000    0.000000    0.385000   
25%    123.000000    0.078960    0.000000    5.130000    0.000000    0.453000   
50%    244.000000    0.261690    0.000000    9.900000    0.000000    0.538000   
75%    377.000000    3.678220   12.500000   18.100000    0.000000    0.631000   
max    506.000000   73.534100  100.000000   27.740000    1.000000    0.871000   

               rm         age         dis         rad         tax     ptratio  \
count  333.000000  333.000000  333.000000  333.000000  333.000000  333.000000   
mean     6.265619   68.226426    3.709934    9.633634  409.279279   18.448048   
std      0.703952   28.133344    1.981123    8.742174  170.841988    2.151821   
min      3.561000    6.000000    1.129600    1.000000  188.000000   12.600000   
25%      5.884000   45.400000    2.122400    4.000000  279.000000   17.400000   
50%      6.202000   76.700000    3.092300    5.000000  330.000000   19.000000   
75%      6.595000   93.800000    5.116700   24.000000  666.000000   20.200000   
max      8.725000  100.000000   10.710300   24.000000  711.000000   21.200000   

            black       lstat        medv  
count  333.000000  333.000000  333.000000  
mean   359.466096   12.515435   22.768769  
std     86.584567    7.067781    9.173468  
min      3.500000    1.730000    5.000000  
25%    376.730000    7.180000   17.400000  
50%    392.050000   10.970000   21.600000  
75%    396.240000   16.420000   25.000000  
max    396.900000   37.970000   50.000000

In [56]:
#checking corrilation between numerical variables
data.corr()

ID      crim        zn     indus      chas       nox        rm  \
ID       1.000000  0.456312 -0.155639  0.421978  0.007958  0.440185 -0.112790   
crim     0.456312  1.000000 -0.210913  0.422228 -0.041195  0.463001 -0.310180   
zn      -0.155639 -0.210913  1.000000 -0.518679 -0.024442 -0.501990  0.328197   
indus    0.421978  0.422228 -0.518679  1.000000  0.037496  0.750087 -0.440365   
chas     0.007958 -0.041195 -0.024442  0.037496  1.000000  0.080275  0.112251   
nox      0.440185  0.463001 -0.501990  0.750087  0.080275  1.000000 -0.338515   
rm      -0.112790 -0.310180  0.328197 -0.440365  0.112251 -0.338515  1.000000   
age      0.257300  0.379034 -0.544513  0.638378  0.068286  0.736000 -0.248573   
dis     -0.356461 -0.397067  0.637142 -0.702327 -0.081834 -0.769364  0.269191   
rad      0.707526  0.666636 -0.303663  0.569779  0.007714  0.612180 -0.272783   
tax      0.686246  0.617081 -0.311180  0.708313 -0.021826  0.670722 -0.356987   
ptratio  0.309838  0.313409 -0.380449  0.391087 -0.125067  0.192513 -0.366927   
black   -0.271619 -0.475796  0.168130 -0.335049  0.062029 -0.369416  0.155202   
lstat    0.281953  0.532077 -0.388112  0.614155 -0.050055  0.598874 -0.615747   
medv    -0.221694 -0.407454  0.344842 -0.473932  0.204390 -0.413054  0.689598   

              age       dis       rad       tax   ptratio     black     lstat  \
ID       0.257300 -0.356461  0.707526  0.686246  0.309838 -0.271619  0.281953   
crim     0.379034 -0.397067  0.666636  0.617081  0.313409 -0.475796  0.532077   
zn      -0.544513  0.637142 -0.303663 -0.311180 -0.380449  0.168130 -0.388112   
indus    0.638378 -0.702327  0.569779  0.708313  0.391087 -0.335049  0.614155   
chas     0.068286 -0.081834  0.007714 -0.021826 -0.125067  0.062029 -0.050055   
nox      0.736000 -0.769364  0.612180  0.670722  0.192513 -0.369416  0.598874   
rm      -0.248573  0.269191 -0.272783 -0.356987 -0.366927  0.155202 -0.615747   
age      1.000000 -0.764208  0.447380  0.511893  0.259293 -0.268054  0.588834   
dis     -0.764208  1.000000 -0.477610 -0.529539 -0.231101  0.284374 -0.505939   
rad      0.447380 -0.477610  1.000000  0.903562  0.470849 -0.406405  0.484568   
tax      0.511893 -0.529539  0.903562  1.000000  0.467437 -0.406477  0.544485   
ptratio  0.259293 -0.231101  0.470849  0.467437  1.000000 -0.164614  0.374802   
black   -0.268054  0.284374 -0.406405 -0.406477 -0.164614  1.000000 -0.356693   
lstat    0.588834 -0.505939  0.484568  0.544485  0.374802 -0.356693  1.000000   
medv    -0.358888  0.249422 -0.352251 -0.448078 -0.481376  0.336660 -0.738600   

             medv  
ID      -0.221694  
crim    -0.407454  
zn       0.344842  
indus   -0.473932  
chas     0.204390  
nox     -0.413054  
rm       0.689598  
age     -0.358888  
dis      0.249422  
rad     -0.352251  
tax     -0.448078  
ptratio -0.481376  
black    0.336660  
lstat   -0.738600  
medv     1.000000

In [57]:
#ID columns does not relevant for our analysis.
data.drop(['ID'], axis = 1, inplace=True)

## Asssigning independent variables and dependent variable to 'X' and 'y'

In [58]:
X = data[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis','rad', 'tax', 'ptratio', 'black', 'lstat']]
y = data['medv']

In [59]:
#Storing columns names
columns_X = ['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis','rad', 'tax', 'ptratio', 'black', 'lstat']
columns_y = ['medv']

In [60]:
X

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
3    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
4    0.08829  12.5   7.87     0  0.524  6.012  66.6  5.5605    5  311   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
328  0.17783   0.0   9.69     0  0.585  5.569  73.5  2.3999    6  391   
329  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
330  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
331  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
332  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio   black  lstat  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       18.7  394.63   2.94  
3       18.7  396.90   5.33  
4       15.2  395.60  12.43  
..       ...     ...    ...  
328     19.2  395.77  15.10  
329     21.0  391.99   9.67  
330     21.0  396.90   9.08  
331     21.0  396.90   5.64  
332     21.0  396.90   7.88  

[333 rows x 13 columns]

In [61]:
y

0      24.0
1      21.6
2      33.4
3      36.2
4      22.9
       ... 
328    17.5
329    22.4
330    20.6
331    23.9
332    11.9
Name: medv, Length: 333, dtype: float64

## Creating " train_test_split " function to split the dataset

In [62]:
import random

def train_test_split(X, y, test_size=0.2, random_state=None):
    
    X=X.values
    # Set random seed
    if random_state:
        random.seed(random_state)

    # Determine number of test samples
    n_samples = len(X)
    n_test = int(test_size * n_samples)

    # Shuffle indices
    indices = list(range(n_samples))
    random.shuffle(indices)

    # Split indices into training and testing sets
    test_indices = indices[:n_test]
    train_indices = indices[n_test:]

    # Split data into training and testing sets
    X_train = [X[i] for i in train_indices]
    X_test = [X[i] for i in test_indices]
    y_train = [y[i] for i in train_indices]
    y_test = [y[i] for i in test_indices]

    return X_train, X_test, y_train, y_test


In [63]:
# Split the data into training and testing sets.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 42)


## Defining " StandardScaler " function to scale the data

In [64]:
import numpy as np

class StandardScaler:
    def __init__(self):
        self.mean = None
        self.std = None
    
    def fit(self, X):
        self.mean = np.mean(X, axis=0)
        self.std = np.std(X, axis=0)
    
    def transform(self, X):
        X_scaled = ((X - self.mean) / self.std)
        return X_scaled
    
    def fit_transform(self, X):
        self.fit(X)
        X_scaled = self.transform(X)
        return X_scaled

In [65]:
# Scaling the dataset
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [66]:
# Converting scaled data to DataFrame
X_train_scaled = pd.DataFrame(X_train,columns = columns_X)
X_test_scaled = pd.DataFrame(X_test,columns = columns_X)
y_train_s = pd.DataFrame(y_train,columns = columns_y)
y_test_s = pd.DataFrame(y_test,columns = columns_y)

In [67]:
X_train_scaled.head()

crim        zn     indus      chas       nox        rm       age  \
0 -0.409068  0.461073 -0.743356 -0.268866 -1.082013  0.263654 -1.966647   
1 -0.271793 -0.495770  1.535723 -0.268866  0.639096  0.035763  1.067527   
2 -0.108293 -0.495770  1.207297 -0.268866  2.841759 -1.456145  0.934983   
3 -0.385859 -0.495770 -0.168966 -0.268866 -0.074317 -0.547411  0.548099   
4 -0.376850 -0.495770  2.070305 -0.268866  0.255636 -0.973469  0.992299   

        dis       rad       tax   ptratio     black     lstat  
0  1.830619 -0.266148 -0.403498  0.319545  0.411127 -0.894929  
1 -0.747264 -0.616205  0.227180  1.306700  0.418166  0.016422  
2 -1.016132 -0.499520  0.026778 -1.748781 -3.081662  0.572404  
3  0.121603 -0.616205 -0.556747 -0.009507  0.347777  0.886219  
4 -1.005671 -0.849577 -1.240474  0.319545 -0.008825  2.165837

In [68]:
X_test_scaled.head()

crim        zn     indus      chas       nox        rm       age  \
0 -0.420359 -0.495770  0.402581 -0.268866 -1.028507  0.351414 -0.605388   
1 -0.417095  1.461409 -1.087422 -0.268866 -1.028507  0.681219 -0.960032   
2  1.441891 -0.495770  0.996877 -0.268866  0.549919 -0.101538  0.720047   
3 -0.418735 -0.495770 -1.115857 -0.268866 -0.823401 -0.612523 -1.508118   
4  0.538176 -0.495770  0.996877 -0.268866  0.398319 -0.967807  1.074691   

        dis       rad       tax   ptratio     black     lstat  
0  1.108256 -0.616205 -0.645160 -1.137685  0.372186 -0.681421  
1  0.015987 -0.499520 -0.002693 -1.513744  0.383766 -0.783160  
2 -0.908102  1.717508  1.576950  0.836626  0.263991  0.138223  
3  0.733169 -0.616205  0.185921 -0.714618  0.254000 -0.311722  
4 -1.157808  1.717508  1.576950  0.836626 -0.515742  2.042604

In [69]:
y_train_s

medv
0    24.4
1    19.6
2    13.1
3    17.8
4    15.7
..    ...
262  37.2
263  24.1
264  20.2
265  23.9
266  18.3

[267 rows x 1 columns]

In [70]:
y_test_s

medv
0   27.1
1   32.0
2   21.4
3   19.3
4   17.2
..   ...
61  11.7
62  19.9
63  21.4
64  10.5
65  35.4

[66 rows x 1 columns]

## Building the LassoRegression algorithm from skratch

In [71]:
import numpy as np

class LassoRegression:
    def __init__(self, alpha, max_iter=1000, tol=1e-4):
        self.alpha = alpha
        self.max_iter = max_iter
        self.tol = tol
        self.coef_ = None
        self.intercept_ = None
    
    def fit(self, X, y):
        m, n = X.shape
        theta = np.zeros(n)
        self.intercept_ = np.mean(y)
        X_centered = X - np.mean(X, axis=0)
        for iteration in range(self.max_iter):
            old_theta = theta.copy()
            for j in range(n):
                X_j = X_centered[:, j]
                y_pred = X_centered.dot(theta) + self.intercept_
                r = y - y_pred + theta[j] * X_j
                z = X_j.dot(X_j)
                if z == 0:
                    theta[j] = 0
                else:
                    theta[j] = np.sign(np.sum(X_j*r)) * max(0, np.abs(np.sum(X_j*r)) - self.alpha) / z
            self.intercept_ = np.mean(y - X_centered.dot(theta))
            if np.sum(np.abs(theta - old_theta)) < self.tol:
                break
        self.coef_ = theta
    
    def predict(self, X):
        return X.dot(self.coef_) + self.intercept_

In [72]:
# Fit the Lasso regression model
lasso = LassoRegression(alpha=1)
lasso.fit(X_train, y_train)

# Make predictions on the testing set
y_pred = lasso.predict(X_test)

In [73]:
y_pred.shape

(66,)

## Performence matrics

### R2_score

In [74]:
import numpy as np

def r2_score(y_test, y_pred):
    sse = np.sum((y_test - y_pred) ** 2)
    sst = np.sum((y_test - np.mean(y_test)) ** 2)
    r2 = 1 - sse / sst
    return r2


In [75]:
r2 = r2_score(y_test, y_pred)
print("R-squared:", r2*100)

R-squared: 77.88090285819405


### Adjusted R2_Score

In [76]:
import numpy as np

def adjusted_r2_score(y_test, y_pred, n_features):
    sse = np.sum((y_test - y_pred) ** 2)
    sst = np.sum((y_test - np.mean(y_test)) ** 2)
    r2 = 1 - sse / sst
    adj_r2 = 1 - ((1 - r2) * (len(y_test) - 1)) / (len(y_test) - n_features - 1)
    return adj_r2


In [77]:
n_features = 1
adj_r2 = adjusted_r2_score(y_test, y_pred, n_features)
print("Adjusted R-squared:", adj_r2*100)


Adjusted R-squared: 77.53529196535332


### Mean Absolute Error

In [78]:
import numpy as np

def mean_absolute_error(y_test, y_pred):
    mae = np.mean(np.abs(y_test - y_pred))
    return mae


In [79]:
mae = mean_absolute_error(y_test, y_pred)
print("Mean absolute error:", mae)

Mean absolute error: 2.8055740193607863


### Mean Squared Error

In [80]:
import numpy as np

def mean_squared_error(y_test, y_pred):
    mse = np.mean((y_test - y_pred) ** 2)
    return mse

In [81]:
mse = mean_squared_error(y_test, y_pred)
print("Mean squared error:", mse)


Mean squared error: 15.243409309059123


### Root Mean Squared Error

In [82]:
import numpy as np

def root_mean_squared_error(y_test, y_pred):
    mse = np.mean((y_test - y_pred) ** 2)
    rmse = np.sqrt(mse)
    return rmse


In [83]:
rmse = root_mean_squared_error(y_test, y_pred)
print("Root mean squared error:", rmse)

Root mean squared error: 3.9042808952557606


# Hyperparameter tuning

## Grid search CV

### Building the "Grid-search-CV" function

In [84]:
import numpy as np

def lasso_regression(X_train, y_train, X_test, y_test, alpha):
    
    lasso = LassoRegression(alpha=alpha)
    lasso.fit(X_train, y_train)
    y_pred = lasso.predict(X_test)
    Mse = np.mean((y_pred - y_test)**2)
    return Mse

def grid_search(X_train, y_train, X_test, y_test, alpha_values):
    
    best_alpha = None
    best_mse = float('inf')
    
    for alpha in alpha_values:
        Mse = lasso_regression(X_train, y_train, X_test, y_test, alpha)
        if Mse < best_mse:
            best_mse = Mse
            best_alpha = alpha
    
    return best_alpha

In [85]:
alpha_values = [0.1, 1, 9, 10,100]
best_alpha = grid_search(X_train, y_train, X_test, y_test, alpha_values)

print("Best alpha:", best_alpha)

Best alpha: 10


In [86]:
# Fit the Lasso regression model
lasso_grid = LassoRegression(alpha=10)
lasso_grid.fit(X_train, y_train)

# Make predictions on the testing set
y_pred_grid = lasso_grid.predict(X_test)

In [87]:
import numpy as np

def r2_score(y_test, y_pred_grid):
    sse = np.sum((y_test - y_pred_grid) ** 2)
    sst = np.sum((y_test - np.mean(y_test)) ** 2)
    r2 = 1 - sse / sst
    return r2


In [88]:
r2_grid = r2_score(y_test, y_pred_grid)
print("R-squared:", r2_grid*100)


R-squared: 78.09998690952604


In [89]:
import numpy as np

def adjusted_r2_score(y_test, y_pred_grid, n_features):
    sse = np.sum((y_test - y_pred_grid) ** 2)
    sst = np.sum((y_test - np.mean(y_test)) ** 2)
    r2 = 1 - sse / sst
    adj_r2 = 1 - ((1 - r2) * (len(y_test) - 1)) / (len(y_test) - n_features - 1)
    return adj_r2


In [90]:
n_features = 1
adj_r2_grid = adjusted_r2_score(y_test, y_pred_grid, n_features)
print("Adjusted R-squared:", adj_r2_grid*100)


Adjusted R-squared: 77.75779920498738


In [91]:
import numpy as np

def mean_absolute_error(y_test, y_pred_grid):
    mae = np.mean(np.abs(y_test - y_pred_grid))
    return mae


In [92]:
mae_grid = mean_absolute_error(y_test, y_pred_grid)
print("Mean absolute error:", mae_grid)

Mean absolute error: 2.8180580734118115


In [93]:
import numpy as np

def mean_squared_error(y_test, y_pred_grid):
    mse = np.mean((y_test - y_pred_grid) ** 2)
    return mse

In [94]:
mse_grid = mean_squared_error(y_test, y_pred_grid)
print("Mean squared error:", mse_grid)

Mean squared error: 15.092427203138149


In [95]:
import numpy as np

def root_mean_squared_error(y_test, y_pred_grid):
    mse = np.mean((y_test - y_pred_grid) ** 2)
    rmse = np.sqrt(mse)
    return rmse


In [96]:
rmse_grid = root_mean_squared_error(y_test, y_pred_grid)
print("Root mean squared error:", rmse_grid)

Root mean squared error: 3.8848973220843495


In [138]:
import pickle
# open a file, where you ant to store the data
file1 = open('lasso1.pkl', 'wb')

# dump information to that file
pickle.dump(lasso, file1)
file1.close()

In [139]:
from sklearn.linear_model import Lasso
sk_model = Lasso(alpha=0.01)

In [140]:
sk_model.fit(X_train, y_train)

Lasso(alpha=0.01)

In [141]:
y_pred = sk_model.predict(X_test)

In [142]:
from sklearn import metrics
print('Lasso')
print('r2_score', "{:.2f}".format(metrics.r2_score(y_test,y_pred)*100))
score = metrics.r2_score(y_test,y_pred)
a_r2 = 1-(1-score)*(len(y_test)-1)/(len(y_test)-X_test.shape[1]-1)
print('Adjusted_r2_score',"{:.2f}".format(a_r2*100))
print('MAE:', "{:.2f}".format(metrics.mean_absolute_error(y_test, y_pred)))
print('MSE:', "{:.2f}".format(metrics.mean_squared_error(y_test, y_pred)))
print('RMSE:', "{:.2f}".format(np.sqrt(metrics.mean_squared_error(y_test, y_pred))))

Lasso
r2_score 77.93
Adjusted_r2_score 72.42
MAE: 2.81
MSE: 15.21
RMSE: 3.90


In [144]:
a = [[1,1,1,1,1,1,1,1,1,1,1,1,1]]
a= np.array(a)

In [145]:
lasso.predict(a)

array([18.68237643])

In [147]:
data.head(1)

crim    zn  indus  chas    nox     rm   age   dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.09    1  296     15.3   

   black  lstat  medv  
0  396.9   4.98  24.0

In [148]:
data.to_csv("data.csv")

## Random-Search-CV

### Creating "Random-Search_CV" function

In [149]:
import numpy as np
import random

def lasso_regression(X_train, y_train, alpha):
    n_samples, n_features = X_train.shape
    w = np.zeros(n_features)
    b = 0

    learning_rate = 0.01
    iterations = 1000

    for i in range(iterations):
        y_pred = np.dot(X_train, w) + b
        mse = np.mean((y_pred - y_train) ** 2)
        dw = (1 / n_samples) * np.dot(X_train.T, (y_pred - y_train))
        db = (1 / n_samples) * np.sum(y_pred - y_train)

        w = w - learning_rate * (dw + alpha * np.sign(w))
        b = b - learning_rate * db

    return w, b

def random_search(X, y, alpha_values, n_iter):
    results = {}

    for i in range(n_iter):
        alpha = random.choice(alpha_values)

        w, b = lasso_regression(X, y, alpha)
        y_pred = np.dot(X, w) + b
        mse = np.mean((y_pred - y) ** 2)

        results[alpha] = mse

    return results

def hyperparameter_tuning(X, y, alpha_values, n_iter):
    results = random_search(X, y, alpha_values, n_iter)
    best_alpha = min(results, key=results.get)
    return best_alpha


In [109]:
alpha_values = [0.01,0.1, 1, 10,100]
n_iter = 5

best_alpha = hyperparameter_tuning(X_train, y_train, alpha_values, n_iter)
print("Best alpha:", best_alpha)

Best alpha: 0.1


In [110]:
# Fit the Lasso regression model
lasso_rand = LassoRegression(alpha=0.1)
lasso_rand.fit(X_train, y_train)

# Make predictions on the testing set
y_pred_rand = lasso_rand.predict(X_test)

In [111]:
import numpy as np

def r2_score(y_test, y_pred_rand):
    sse = np.sum((y_test - y_pred_rand) ** 2)
    sst = np.sum((y_test - np.mean(y_test)) ** 2)
    r2 = 1 - sse / sst
    return r2


In [112]:
r2_rand = r2_score(y_test, y_pred_rand)
print("R-squared:", r2_rand*100)

R-squared: 77.85095545532522


In [113]:
import numpy as np

def adjusted_r2_score(y_test, y_pred_rand, n_features):
    sse = np.sum((y_test - y_pred_rand) ** 2)
    sst = np.sum((y_test - np.mean(y_test)) ** 2)
    r2 = 1 - sse / sst
    adj_r2 = 1 - ((1 - r2) * (len(y_test) - 1)) / (len(y_test) - n_features - 1)
    return adj_r2


In [114]:
n_features = 1
adj_r2_rand = adjusted_r2_score(y_test, y_pred_rand, n_features)
print("Adjusted R-squared:", adj_r2_rand*100)


Adjusted R-squared: 77.50487663431468


In [115]:
import numpy as np

def mean_absolute_error(y_test, y_pred_rand):
    mae = np.mean(np.abs(y_test - y_pred_rand))
    return mae


In [116]:
mae_rand = mean_absolute_error(y_test, y_pred_rand)
print("Mean absolute error:", mae_rand)

Mean absolute error: 2.8058576329674367


In [117]:
import numpy as np

def mean_squared_error(y_test, y_pred_rand):
    mse = np.mean((y_test - y_pred_rand) ** 2)
    return mse

In [118]:
mse_rand = mean_squared_error(y_test, y_pred_rand)
print("Mean squared error:", mse_rand)

Mean squared error: 15.264047607121022


In [119]:
import numpy as np

def root_mean_squared_error(y_test, y_pred_rand):
    mse = np.mean((y_test - y_pred_rand) ** 2)
    rmse = np.sqrt(mse)
    return rmse


In [120]:
rmse_rand = root_mean_squared_error(y_test, y_pred_rand)
print("Root mean squared error:", rmse_rand)

Root mean squared error: 3.9069230357304225


## K-Fold-Cross-Validation

### Creating "K-Fold-Cross-Validation" function

In [121]:
import numpy as np

def lasso_regression(X_train, y_train, alpha):
    n_samples, n_features = X_train.shape
    w = np.zeros(n_features)
    b = 0

    learning_rate = 0.01
    iterations = 1000

    for i in range(iterations):
        y_pred = np.dot(X_train, w) + b
        mse = np.mean((y_pred - y_train) ** 2)
        dw = (1 / n_samples) * np.dot(X_train.T, (y_pred - y_train))
        db = (1 / n_samples) * np.sum(y_pred - y_train)

        w = w - learning_rate * (dw + alpha * np.sign(w))
        b = b - learning_rate * db

    return w, b

def k_fold_cross_validation(X, y, k, alpha_values):
    fold_size = len(X) // k
    results = {}

    for alpha in alpha_values:
        mse_sum = 0
        for i in range(k):
            val_start = i * fold_size
            val_end = (i + 1) * fold_size

            X_val = X[val_start:val_end]
            y_val = y[val_start:val_end]

            X_train = np.concatenate([X[:val_start], X[val_end:]])
            y_train = np.concatenate([y[:val_start], y[val_end:]])

            w, b = lasso_regression(X_train, y_train, alpha)
            y_pred = np.dot(X_val, w) + b
            mse_sum += np.mean((y_pred - y_val) ** 2)

        results[alpha] = mse_sum / k

    return results

def hyperparameter_tuning(X, y, k, alpha_values):
    results = k_fold_cross_validation(X, y, k, alpha_values)
    best_alpha = min(results, key=results.get)
    return best_alpha


In [122]:
k = 3
alpha_values = [0.01, 0.1,0.5, 1, 5, 10, 100]

best_alpha = hyperparameter_tuning(X_train, y_train, k, alpha_values)
print("Best alpha:", best_alpha)

Best alpha: 0.01


In [123]:
# Fit the Lasso regression model
lasso_kfold = LassoRegression(alpha=0.01)
lasso_kfold.fit(X_train, y_train)

# Make predictions on the testing set
y_pred_kfold = lasso_kfold.predict(X_test)

In [124]:
import numpy as np

def r2_score(y_test, y_pred_kfold):
    sse = np.sum((y_test - y_pred_kfold) ** 2)
    sst = np.sum((y_test - np.mean(y_test)) ** 2)
    r2 = 1 - sse / sst
    return r2


In [125]:
r2_kfold = r2_score(y_test, y_pred_kfold)
print("R-squared:", r2_kfold*100)

R-squared: 77.84785902965612


In [126]:
import numpy as np

def adjusted_r2_score(y_test, y_pred_kfold, n_features):
    sse = np.sum((y_test - y_pred_kfold) ** 2)
    sst = np.sum((y_test - np.mean(y_test)) ** 2)
    r2 = 1 - sse / sst
    adj_r2 = 1 - ((1 - r2) * (len(y_test) - 1)) / (len(y_test) - n_features - 1)
    return adj_r2


In [127]:
n_features = 1
adj_r2_kfold = adjusted_r2_score(y_test, y_pred_kfold, n_features)
print("Adjusted R-squared:", adj_r2_kfold*100)


Adjusted R-squared: 77.50173182699449


In [128]:
import numpy as np

def mean_absolute_error(y_test, y_pred_kfold):
    mae = np.mean(np.abs(y_test - y_pred_kfold))
    return mae


In [129]:
mae_kfold = mean_absolute_error(y_test, y_pred_kfold)
print("Mean absolute error:", mae_kfold)

Mean absolute error: 2.8059474775537003


In [130]:
import numpy as np

def mean_squared_error(y_test, y_pred_kfold):
    mse = np.mean((y_test - y_pred_kfold) ** 2)
    return mse

In [131]:
mse_kfold = mean_squared_error(y_test, y_pred_kfold)
print("Mean squared error:", mse_kfold)


Mean squared error: 15.266181513562444


In [132]:
import numpy as np

def root_mean_squared_error(y_test, y_pred_kfold):
    mse = np.mean((y_test - y_pred_kfold) ** 2)
    rmse = np.sqrt(mse)
    return rmse


In [133]:
rmse_kfold = root_mean_squared_error(y_test, y_pred_kfold)
print("Root mean squared error:", rmse_kfold)

Root mean squared error: 3.9071961191578857


# Conclusion
## Grid Search has given the best R2_score : 78.09
## And best MSE, RMSE : 15.09, 3.88

#### ****

# ****** 

## Performance matrics

## Lasso_Regression

In [134]:
print("R-squared for lasso:", "{:.2f}".format(r2*100))
print("Adjusted R-squared for lasso:", "{:.2f}".format(adj_r2*100))
print("Mean absolute error for lasso:", "{:.2f}".format(mae))
print("Mean squared error for lasso:", "{:.2f}".format(mse))
print("Root mean squared error for lasso:", "{:.2f}".format(rmse))


R-squared for lasso: 77.88
Adjusted R-squared for lasso: 77.54
Mean absolute error for lasso: 2.81
Mean squared error for lasso: 15.24
Root mean squared error for lasso: 3.90


## Grid-Search-CV

In [135]:
print("R-squared for grid search cv:", "{:.2f}".format(r2_grid*100))
print("Adjusted R-squared for grid search cv:", "{:.2f}".format(adj_r2_grid*100))
print("Mean absolute error for grid search cv:", "{:.2f}".format(mae_grid))
print("Mean squared error for grid search cv:", "{:.2f}".format(mse_grid))
print("Root mean squared error for grid search cv:", "{:.2f}".format(rmse_grid))


R-squared for grid search cv: 78.10
Adjusted R-squared for grid search cv: 77.76
Mean absolute error for grid search cv: 2.82
Mean squared error for grid search cv: 15.09
Root mean squared error for grid search cv: 3.88


## Random-Search-CV

In [136]:
print("R-squared for random search cv:", "{:.2f}".format(r2_rand*100))
print("Adjusted R-squared for random search cv:", "{:.2f}".format(adj_r2_rand*100))
print("Mean absolute error for random search cv:", "{:.2f}".format(mae_rand))
print("Mean squared error for random search cv:", "{:.2f}".format(mse_rand))
print("Root mean squared error for random search cv:", "{:.2f}".format(rmse_rand))


R-squared for random search cv: 77.85
Adjusted R-squared for random search cv: 77.50
Mean absolute error for random search cv: 2.81
Mean squared error for random search cv: 15.26
Root mean squared error for random search cv: 3.91


## Kfold-Cross-Validation

In [137]:
print("R-squared for Kfold cross validation:", "{:.2f}".format(r2_kfold*100))
print("Adjusted R-squared for Kfold cross validation:", "{:.2f}".format(adj_r2_kfold*100))
print("Mean absolute error for Kfold cross validation:", "{:.2f}".format(mae_kfold))
print("Mean squared error for Kfold cross validation:", "{:.2f}".format(mse_kfold))
print("Root mean squared error for Kfold cross validation:", "{:.2f}".format(rmse_kfold))

R-squared for Kfold cross validation: 77.85
Adjusted R-squared for Kfold cross validation: 77.50
Mean absolute error for Kfold cross validation: 2.81
Mean squared error for Kfold cross validation: 15.27
Root mean squared error for Kfold cross validation: 3.91


#####    Reddappa M R